src/storage/src/source/mod.rs

```diff
@@ -952,162 +930,191 @@ where
+            //      I checked and on main, we pass empty start_offsets in tests like materializations.td
+            let start_offsets = vec![];
+
+            let source_reader = S::new(
                 name.clone(),
                 id,
                 worker_id,
                 worker_count,
-                scope.sync_activator_for(&info.address[..]),
+                sync_activator,
                 source_connector.clone(),
                 aws_external_id.clone(),
-                restored_offsets,
+                start_offsets,
                 encoding,
-                base_metrics.clone(),
-            ) {
-                Ok(source_reader) => Some(source_reader.into_stream(timestamp_frequency)),
+                base_metrics,
+            );
+            let source_stream = match source_reader {
+                Ok(s) => s.into_stream(timestamp_frequency).fuse(),
                 Err(e) => {
                     error!("Failed to create source: {}", e);
-                    None
-                }
-            }
-        };
-
-        move |cap, durability_cap, output| {
-            // First check that the source was successfully created
-            let source_reader = match &mut source_reader {
-                Some(source_reader) => source_reader,
-                None => {
-                    return SourceStatus::Done;
+                    return;
                 }
             };
 
-            // Check that we have a valid list of timestamp bindings.
-            let timestamp_histories = match &mut timestamp_histories {
-                Some(histories) => histories,
-                None => {
-                    error!("Source missing list of timestamp bindings");
-                    return SourceStatus::Done;
+            tokio::pin!(source_stream);
+
+            let mut timestamp_interval = tokio::time::interval(timestamp_frequency);
+            let mut pending_messages = vec![];
+            loop {
+                tokio::select! {
+                    // N.B. This branch is cancel-safe because `next` only borrows the underlying stream.
+                    item = source_stream.next() => {
+                        match item {
+                            Some(Ok(message)) => pending_messages.push(Ok(message)),
+                            // TODO: make errors definite
+                            Some(Err(e)) => pending_messages.push(Err(e)),
+                            None => {},
+                        }
+                    }
+                    // It's time to timestamp a batch
+                    _ = timestamp_interval.tick() => {
+                        let mut max_offsets = HashMap::new();
+                        for message in pending_messages.iter().filter_map(|m| m.as_ref().ok()) {
+                            let entry = max_offsets.entry(message.partition.clone()).or_default();
+                            *entry = std::cmp::max(*entry, message.offset);
```

Comment:
Since we're going through all the pending messages here it looks like this logic actually belongs above when we receive each message, in the handler of `Some(Ok(message))`

Reply:
I thought about this some and I think leaving it here is preferable.  The big reason is that if we move it above to have the same scope as `pending_messages` we have to use it and then remember to clear it each time -- making the code a little more complex to follow for, I think, pretty small perf benefit (presumably that's why you're suggesting?) that is certainly not the dominating factor.

The alternative to that is having `max_offsets` stick around and we don't send the reclocker what we believe are new offsets but rather just always max offsets.  That seems less desirable because then we need to add some sort of extra argument (perhaps just even a bool `pending_messages.is_empty()`) to the call to the reclocker so that we can differentiate the following:
1) We're just asking for the upper to advance so that the system can move forward -- in a case where there are no new bindings to mint
2) We're matching to offsets that already exist and don't actually require the system to move forward

We want to differentiate these because I think it adds extra complexity to add a constraint that we want every call to the reclocker to move forward the upper -- that isn't a reasonable constraint in the case of, e.g., reading bindings in after restart.

I also think it make the implementation of the reclocker more complex to have to handle the case where we're getting bindings that we know we've already seen and don't actually care about the timestamps for -- because that makes the computation of the progress capability more complex.  This is because we're unable to differentiate between a matched offset that we _do_ want to hold the progress back for (e.g. a message we've just assigned a binding to and are processing for the first time) and one we don't (a message that we've previously processed but it hanging around in the max_offsets hashmap)

I know that's a very long-winded answer but I think the way it's current written makes it more readable than the way it would need to look if the `max_offset` hashmap were to be computed as we read in messages.